File: runme/mcgen/local/datan/solsrc/s3lsq.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

extern double CBFUNC powerl(double const P_T *x, integer const P_T *nr,
                            double const P_T *t);

/* Table of constant values */

static integer c__1 = 1;
static double c_dp5 = .5;

main(void)
{
    /* Local variables */
    static integer i, n, ipl, lcapt, list[2], npos, ltx, lty, nws, npl, nstep,
                   nr = 2;
    static double a[200] /* was [100][2] */, cx[4] /* was [2][2] */,
                  dellog[100], deltat, datcov[100], deltay[100], dt, dx1, dx2,
                  p, r, t[100], x[2], y[100], sigma, rho,
                  scrat[4] /* was [2][2] */, datsx[100], t0, tlog[100],
                  ylog[100], xpl[200], ypl[200];
    static char capt[75], tx[75], ty[75];
    static logical ok;
    /* System generated locals */
    integer i__1;
/* identify program to user */
    printf("%s\n", "Program S3LSQ generates data");
    printf("%s\n\n", "and fits a power law (nonlinear case).");
/* ask for input */
    printf("%s\n", "Enter number n of data points (3 <= n <= 100):");
    printf("%s", "> ");
    scanf("%li", &n);
    printf("%s\n", "Enter first value T0 of controlled variable:");
    printf("%s", "> ");
    scanf("%lf", &t0);
    printf("%s\n", "Enter step width DELTAT of controlled variable:");
    printf("%s", "> ");
    scanf("%lf", &deltat);
    printf("%s\n", "Enter coefficient X(1):");
    printf("%s", "> ");
    scanf("%lf", &x[0]);
    printf("%s\n", "Enter exponent X(2):");
    printf("%s", "> ");
    scanf("%lf", &x[1]);
    printf("%s\n", "Enter size of measurement errors (>0.):");
    printf("%s", "> ");
    scanf("%lf", &sigma);
/* generate data points corresponding to power law */
    rnstnr_(deltay, &n);
    i__1 = n;
    for (i = 1; i <= i__1; ++i) {
        t[i - 1] = t0 + (double) (i - 1) * deltat;
        y[i - 1] = powerl(x, &nr, &t[i - 1]);
        y[i - 1] += deltay[i - 1] * sigma;
        deltay[i - 1] = sigma;
    }
/* find first approximation of unknowns by straight line in log-log plot */
    npos = 0;
    i__1 = n;
    for (i = 1; i <= i__1; ++i) {
        if (t[i - 1] > 0. && y[i - 1] > 0.) {
            ++npos;
            tlog[npos - 1] = log(t[i - 1]);
            ylog[npos - 1] = log(y[i - 1]);
            dellog[npos - 1] = 1.;
        }
    }
    lsqpol_(tlog, ylog, dellog, &npos, &nr, x, cx, &r, a, scrat, &ok);
    x[0] = exp(x[0]);
/* perform fit */
    nstep = 100;
    lsqnon_(powerl, t, y, deltay, &n, &nr, &nr, list, x, cx, &r, a,
           scrat, &nstep);
    if (n != nr) {
        i__1 = n - nr;
        p = 1. - scchi2_(&r, &i__1);
        dx1 = sqrt(cx[0]);
        dx2 = sqrt(cx[3]);
        rho = cx[2] / (dx1 * dx2);
    }
/* prepare graphics */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    (void)getchar();
    memset(tx, ' ', 75);
    memset(ty, ' ', 75);
    memset(capt, ' ', 75);
    strncpy(tx, "t", 1);
    strncpy(ty, "y", 1);
    ltx = 1;
    lty = 1;
    sprintf(capt, "x_1#=%6.2lf, x_2#=%6.2lf", x[0], x[1]);
    lcapt = strlen(capt);
    if (n != nr) {
        sprintf(capt + 24, ", &D@x_1#=%6.2lf, &D@x_2#=%6.2lf, &r@=%6.2lf",
                dx1, dx2, rho);
        lcapt = strlen(capt);
    }
/* curve corresponding to solution */
    npl = 200;
    dt = (double) (n - 1) / (double) (npl - 1) * deltat;
    i__1 = npl;
    for (ipl = 1; ipl <= i__1; ++ipl) {
        xpl[ipl - 1] = t0 + dt * (double) ipl;
        ypl[ipl - 1] = powerl(x, &nr, &xpl[ipl - 1]);
    }
/* prepare data points for graphical presentation */
    i__1 = n;
    for (i = 1; i <= i__1; ++i) {
        datsx[i - 1] = 0.;
        datcov[i - 1] = 0.;
    }
/* graphical output */
    grdtcv_(xpl, ypl, &npl, &c__1, &c_dp5, t, y, datsx, deltay, datcov, &n,
           tx, &ltx, ty, &lty, capt, &lcapt, &nws);
    return 0;
} /* main */

/* --------------------------------------------- */
double CBFUNC powerl(double const P_T *x, integer const P_T *nr,
                    double const P_T *t)
{
    /* System generated locals */
    double ret_val;

    /* Parameter adjustments */
    --x;

    /* Function Body */
    ret_val = x[1] * pow(*t, x[2]);
    return ret_val;
} /* powerl */

File: runme/mcgen/local/datan/solsrc/s3lsq.f
      PROGRAM S3LSQ
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(MAXNR=2,MAXN=100)
      DIMENSION T(MAXN),DELTAY(MAXN),CX(MAXNR,MAXNR),A(MAXN,MAXNR)
      DIMENSION Y(MAXN),X(MAXNR),SCRAT(MAXNR,MAXNR)
      DIMENSION TLOG(MAXN),YLOG(MAXN),DELLOG(MAXN)
      CHARACTER*75 TX,TY,CAPT
      DIMENSION DATSX(MAXN),DATCOV(MAXN)
      DIMENSION XPL(200),YPL(200)
      DIMENSION LIST(MAXNR)
      LOGICAL OK
      EXTERNAL POWERL
      DATA NR /2/
C identify program to user
      WRITE(*,*)' Program S3LSQ generates data'
      WRITE(*,*)' and fits a power law (nonlinear case)'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter number n of data points (3 <= n <= 100)'
      WRITE(*,*)' >'
      READ(*,*)N
      WRITE(*,*)' Enter first value T0 of controlled variable'
      WRITE(*,*)' >'
      READ(*,*)T0
      WRITE(*,*)' Enter step width DELTAT of controlled variable'
      WRITE(*,*)' >'
      READ(*,*)DELTAT
      WRITE(*,*)' Enter coefficient X(1)'
      WRITE(*,*)' >'
      READ(*,*)X(1)
      WRITE(*,*)' Enter exponent X(2)'
      WRITE(*,*)' >'
      READ(*,*)X(2)
      WRITE(*,*)' Enter size of measurement errors (>0.)'
      WRITE(*,*)' >'
      READ(*,*)SIGMA
C generate data points corresponding to power law
      CALL RNSTNR(DELTAY,N)
      DO 10 I=1,N
        T(I)=T0+DBLE(I-1)*DELTAT
        Y(I)=POWERL(X,NR,T(I))
        Y(I)=Y(I)+DELTAY(I)*SIGMA
        DELTAY(I)=SIGMA
10    CONTINUE
C find first approximation of unknowns by straight line in log-log plot
      NPOS=0
      DO 12 I=1,N
        IF(T(I).GT.0.D0 .AND. Y(I).GT.0.D0) THEN
          NPOS=NPOS+1
          TLOG(NPOS)=LOG(T(I))
          YLOG(NPOS)=LOG(Y(I))
          DELLOG(NPOS)=1.D0
        END IF
12    CONTINUE
      CALL LSQPOL(TLOG,YLOG,DELLOG,NPOS,NR,X,CX,R,A,SCRAT,OK)
      X(1)=EXP(X(1))
C perform fit
      NSTEP=100
      CALL LSQNON(POWERL,T,Y,DELTAY,N,NR,NR,LIST,X,CX,R,A,SCRAT,
     +NSTEP)
      IF(N.NE.NR) THEN
        P=1.D0-SCCHI2(R,N-NR)
        DX1=SQRT(CX(1,1))
        DX2=SQRT(CX(2,2))
        RHO=CX(1,2)/(DX1*DX2)
      END IF
C prepare graphics
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
      TX='t'
      TY='y'
      LTX=1
      LTY=1
      CAPT(1:24)='x_1#=******, x_2#=******'
      WRITE(CAPT(6:11),'(F6.2)')X(1)
      WRITE(CAPT(19:24),'(F6.2)')X(2)
      LCAPT=24
      IF(N.NE.NR) THEN
        CAPT(25:69)=
     *', &D@x_1#=******, &D@x_2#=******, &r@=******'
        WRITE(CAPT(35:40),'(F6.2)')DX1
        WRITE(CAPT(51:56),'(F6.2)')DX2
        WRITE(CAPT(63:68),'(F6.2)')RHO
        LCAPT=68
      END IF
C curve corresponding to solution
      NPL=200
      DT=(DBLE(N-1)/DBLE(NPL-1))*DELTAT
      DO 20 IPL=1,NPL
        XPL(IPL)=T0+DT*DBLE(IPL)
        YPL(IPL)=POWERL(X,NR,XPL(IPL))
20    CONTINUE
C prepare data points for graphical presentation
      DO 30 I=1,N
        DATSX(I)=0.D0
        DATCOV(I)=0.D0
30    CONTINUE
C graphical output
      CALL GRDTCV(XPL,YPL,NPL,1,.5D0,T,Y,DATSX,DELTAY,DATCOV,N,
     +TX,LTX,TY,LTY,CAPT,LCAPT,NWS)
      END
C---------------------------------------------
      DOUBLE PRECISION FUNCTION POWERL(X,NR,T)
      DOUBLE PRECISION X,T
      DIMENSION X(*)
      POWERL=X(1)*T**X(2)
      END
